cmd/csi-external-health-monitor-controller/main.go

```diff
@@ -0,0 +1,303 @@
+	resync        = flag.Duration("resync", 10*time.Minute, "Resync interval of the controller.")
+	csiAddress    = flag.String("csi-address", "/run/csi/socket", "Address of the CSI driver socket.")
+	showVersion   = flag.Bool("version", false, "Show version.")
+	timeout       = flag.Duration("timeout", 15*time.Second, "Timeout for waiting for attaching or detaching the volume.")
+	workerThreads = flag.Uint("worker-threads", 10, "Number of pv monitor worker threads")
+
+	enableLeaderElection    = flag.Bool("leader-election", false, "Enable leader election.")
+	leaderElectionNamespace = flag.String("leader-election-namespace", "", "Namespace where the leader election resource lives. Defaults to the pod namespace if not set.")
+
+	metricsAddress = flag.String("metrics-address", "", "The TCP network address where the prometheus metrics endpoint will listen (example: `:8080`). The default is empty string, which means metrics endpoint is disabled.")
+	metricsPath    = flag.String("metrics-path", "/metrics", "The HTTP path where prometheus metrics will be exposed. Default is `/metrics`.")
+)
+
+var (
+	version = "unknown"
+)
+
+func main() {
+	klog.InitFlags(nil)
+	flag.Set("logtostderr", "true")
+	flag.Parse()
+
+	if *showVersion {
+		fmt.Println(os.Args[0], version)
+		return
+	}
+	klog.Infof("Version: %s", version)
+
+	// Create the client config. Use kubeconfig if given, otherwise assume in-cluster.
+	config, err := buildConfig(*kubeconfig)
+	if err != nil {
+		klog.Error(err.Error())
+		os.Exit(1)
+	}
+
+	if *workerThreads == 0 {
+		klog.Error("option -worker-threads must be greater than zero")
+		os.Exit(1)
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		klog.Error(err.Error())
+		os.Exit(1)
+	}
+
+	factory := informers.NewSharedInformerFactory(clientset, *resync)
+
+	metricsManager := metrics.NewCSIMetricsManager("" /* driverName */)
+
+	// Connect to CSI.
+	csiConn, err := connection.Connect(*csiAddress, metricsManager, connection.OnConnectionLoss(connection.ExitOnConnectionLoss()))
+	if err != nil {
+		klog.Error(err.Error())
+		os.Exit(1)
+	}
+
+	err = rpc.ProbeForever(csiConn, *timeout)
+	if err != nil {
+		klog.Error(err.Error())
+		os.Exit(1)
+	}
+
+	// Find driver name.
+	ctx, cancel := context.WithTimeout(context.Background(), csiTimeout)
+	defer cancel()
+	storageDriver, err := rpc.GetDriverName(ctx, csiConn)
+	if err != nil {
+		klog.Error(err.Error())
+		os.Exit(1)
+	}
+	klog.V(2).Infof("CSI driver name: %q", storageDriver)
+	metricsManager.SetDriverName(storageDriver)
+	metricsManager.StartMetricsEndpoint(*metricsAddress, *metricsPath)
+
+	supportsService, err := supportsPluginControllerService(ctx, csiConn)
+	if err != nil {
+		klog.Error(err.Error())
+		os.Exit(1)
+	}
+	if !supportsService {
+		klog.V(2).Infof("CSI driver does not support Plugin Controller Service, exiting")
+		os.Exit(1)
+	}
+
+	supportGetVolume, err := supportGetVolume(ctx, csiConn)
```

Comment:
We should check LIST_VOLUMES support first.  If both LIST_VOLUMES and VOLUME_CONDITION are supported, we should use that to retrieve volume health.  This will have much better performance than making individual calls to GetVolume.

If LIST_VOLUMES is not support, then we check the combination of GET_VOLUME and VOLUME_CONDITION.